Add more instructions (#4413)
dotnet/dotnet-docker: samples/enable-globalization.md

# Enabling (or disabling) globalization functionality

> .NET includes various [globalization](https://learn.microsoft.com/en-us/dotnet/core/extensions/globalization-and-localization) capabilities, including support for processing natural language text, calendars, currency, and timezones. The .NET implementations for these capabilities comes primarily from operating system libraries, such as [International Components for Unicode (ICU)](https://icu.unicode.org/) and [tzdata](https://en.wikipedia.org/wiki/Tz_database). In some cases, this libraries and their accompanying databases are always available and in other cases it is desired that they be absent because they may be considered prohibitively large.

Reviewer: ```suggestion
.NET includes various [globalization](https://learn.microsoft.com/dotnet/core/extensions/globalization-and-localization) capabilities, including support for processing natural language text, calendars, currency, and time zones. The .NET implementations for these capabilities comes primarily from operating system libraries, such as [International Components for Unicode (ICU)](https://icu.unicode.org/) and [tzdata](https://wikipedia.org/wiki/Tz_database). In some cases, these libraries and their accompanying databases are always available and in other cases it is desired that they be absent because they may be considered prohibitively large.
```

## ICU

In many scenarios, globalization support with ICU is required, for example, to correctly sort a list of strings (particularly characters outside the [ASCII](https://en.wikipedia.org/wiki/ASCII) range). Other applications may not be oriented around natural language or other global concepts, and would prefer to optimize for container size. .NET container images have been curated to offer options for both of these scenarios.

.NET container images that include ICU:

- Alpine `sdk` images
- Debian images
- Ubuntu images

.NET container images that do not include ICU:

- Alpine `aspnet`, `monitor`, `runtime`, `runtime-deps` images
- Ubuntu chiseled images

Images that do not include ICU enable [Globalization Invariant Mode](https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-invariant-mode.md), which provides more basic globalization behaviors in absence of using ICU.

Some users want to add ICU to one of the image types that doesn't include it. It is counter-productive to remove ICU from an image that already includes it.

### Alpine images

> ICU can be added to a .NET Alpine image by adding the following statements to the final stage within a `Dockerfile`, as demonstrated in [Dockerfile.alpine-icu](aspnetapp/Dockerfile.alpine-icu). This Dockerfile fragment adds and configures ICU and disables Globalization invariant mode.

Reviewer: ```suggestion
ICU can be added to a .NET Alpine image by adding the following instructions to the final stage within a `Dockerfile`, as demonstrated in [Dockerfile.alpine-icu](aspnetapp/Dockerfile.alpine-icu). This Dockerfile fragment adds and configures ICU and disables Globalization invariant mode.
```

```Dockerfile
ENV \
    DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=false \
    LC_ALL=en_US.UTF-8 \
    LANG=en_US.UTF-8
RUN apk add --no-cache \
    icu-data-full \
    icu-libs
```

### Ubuntu chiseled images

ICU can be added to an Ubuntu chiseled image, as demonstrated by https://github.com/ubuntu-rocks/dotnet/issues/21.

## Tzdata

> Tzdata provides data about timezones. It is needed in applications that deal with time.

Reviewer: ```suggestion
Tzdata provides data about time zones. It is needed in applications that deal with time.
```

.NET container images that include tzdata:

> - Debian images

Reviewer: Yes, but not as a matter of policy. If Debian decided to remove it, we wouldn't explicitly add it. Should we call that out?

Author: Fixed.

Without `tzdata` installed, `DateTime.UtcNow` and `DateTime.Now` will return the same value, `TimeZoneInfo.Local` will return a value for UTC, and the call to `TimeZoneInfo.FindSystemTimeZoneById` will fail resulting in an exception.

.NET container images do not install `tzdata`, however, Debian images contain it (at the time of writing). If you rely on this library, you need to install it.

### Alpine

Install `tzdata`:

```bash
apk add tzdata
```

### Ubuntu

Install `tzdata`:

```bash
apt update && DEBIAN_FRONTEND=noninteractive && apt install -y tzdata
```

You may need to go through the interactive experience once to find the appropriate `TZ` string. It will be recorded in `/etc/timezone`.

### Launching a container with timezone information

The best practice is to pass timezone information into a container via environment variable.

```bash
$ cat /etc/timezone
America/Los_Angeles
docker run --rm -it -e TZ=$(cat /etc/timezone) app
```

> This approach enables a container image to be launched matching the host, as opposed to setting the information as part of `docker build`.

Reviewer: > enables a container image to be launched matching the host

Is that a common practice? In the world of cloud computing, what time zone the host is in seems irrelevant. I did a [quick search](https://grep.app/search?current=6&q=docker%20run.%2ATZ%3D&regexp=true) and couldn't find any examples that use the host time zone. All are either hardcoding it or have it set to an environment variable that can be set dynamically. That approach seems better to me.
